### ISSUE TITLE
Question about the "Try to fix MSVC build" commit (781a4e667d)

### ISSUE THREAD

**Issue by hkarel** [Tencent/rapidjson#2033]
```cpp
#if RAPIDJSON_HAS_CXX11
            // Use thread-local storage to prevent races between threads.
            // Use static buffer and placement-new to prevent destruction, with
            // alignas() to ensure proper alignment.
            alignas(GenericValue) thread_local static char buffer[sizeof(GenericValue)];
            return *new (buffer) GenericValue();
#elif defined(_MSC_VER) && _MSC_VER < 1900
            // There's no way to solve both thread locality and proper alignment
            // simultaneously.
            __declspec(thread) static char buffer[sizeof(GenericValue)];
            return *new (buffer) GenericValue();
#elif defined(__GNUC__) || defined(__clang__)
            // This will generate -Wexit-time-destructors in clang, but that's
            // better than having under-alignment.
            __thread static GenericValue buffer;
            return buffer;
#else
            // Don't know what compiler this is, so don't know how to ensure
            // thread-locality.
            static GenericValue buffer;
            return buffer;
#endif
```
In the above code, the functional behavior for
```cpp
1)
alignas(GenericValue) thread_local static char buffer[sizeof(GenericValue)];
return *new (buffer) GenericValue();
```
and for 
```cpp
2)
__thread static GenericValue buffer;
return buffer;
```
will be different.

In the first case, the default constructor will be called every time the specified code is accessed. In fact, this will mean nulling/modifying an already existing object. In the commit, they wanted to get rid of a single call to the destructor, but in the end they added a multiple call to the default constructor, in fact, with undefined behavior.
To demonstrate, I created a small example:
```cpp
struct A
{
    int a;
    A() {
        a = 10;
        std::cout << "ctor\n";
    }
    ~A() {
        a = 0;
        std::cout << "dtor\n";
    }
};

A& func1()
{
    static A aa;
    return aa;
}
A& func2()
{
    alignas(A) thread_local static char buffer[sizeof(A)];
    return *new (buffer) A();
}
A& func3()
{
    alignas(A) thread_local static char buffer[sizeof(A)];
    static bool first {true};
    if (first) {
        new (buffer) A();
        first = false;
    }
    return *((A*)buffer);
}

int main(int argc, char* argv[])
{
    func3();
    func3();
    func3();
    return 0;
}

```
Call the function func1(), func2() func3() several times in a row, respectively. Look at the console output.
If I'm wrong in my assumption - tell me where I'm wrong.

**Comment by pkasting** [comment]
It looks from https://stackoverflow.com/questions/27552466/c-is-constructing-object-twice-using-placement-new-undefined-behaviour like the current behavior is not UB as long as we do not rely on side effects of ~GenericValue():

> A program may end the lifetime of any object by reusing the storage which the object occupies or by explicitly calling the destructor for an object of a class type with a non-trivial destructor. For an object of a class type with a non-trivial destructor, the program is not required to call the destructor explicitly before the storage which the object occupies is reused or released; however, if there is no explicit call to the destructor or if a delete-expression (5.3.5) is not used to release the storage, the destructor shall not be implicitly called and any program that depends on the side effects produced by the destructor has undefined behavior.

**Comment by hkarel** [comment]
1. Now in the project, depending on the build conditions, there may be two code options with different behavior. How correct is this?
2. Class GenericValue is large enough and I'm not the author of it, so I can't say for sure that calling the placement constructor again will result in undefined behavior. I can only assume it. The class destructor contains code to deallocate resources, so calling the placement constructor multiple times can cause a memory leak.
3. In the question cited (https://stackoverflow.com/questions/27552466/c-is-constructing-object-twice-using-placement-new-undefined-behaviour), there will be no undefined behavior just because the primary constructor (the default constructor) initializes the parameters to zero, and the placement constructor is only called once. In case GenericValue::operator[] this may not be the case. Also in the above link, the idea is repeated several times that repeated call the placement constructor is a bad idea.
4. The code discussed here was added a long time ago (in 2015), commit 74c8dc. If in such a long time it did not lead to problems, then it is called only once or not at all.
5. I would like to hear a response from the developers of the library how correct the discussed code is. Until then, in my clone of the repository, I will prefer to use the albeit less productive, but traditional approach for static variables.
```cpp
thread_local static GenericValue* NullValue {new GenericValue};
return *NullValue;
```

**Comment by jaepil** [comment]
```c++
thread_local static GenericValue* NullValue {new GenericValue};
return *NullValue;
```

The code above will cause a memory leak when your program creates a new thread and execute this path. Creating temporary, short-living threads on-demand at runtime is common, so I don't think it is an ideal solution.

```c++
alignas(A) thread_local static char buffer[sizeof(A)];
return *new (buffer) A();
```

The code above is okay because the default constructor of `GenericValue` does nothing other than initializing some flags to make it `null`. It doesn't dynamically allocate resources, so there's nothing to do in the destructor before calling the default constructor again. It's not a UB in this case. It's more like calling `reset()` or `clear()` function if `GenericValue` provide them.

**Comment by jaepil** [comment]
However, I think calling the constructor multiple times can be seen as a hack. I agree that there's room to improve the code further in some ways, such as making a wrapper class or using a flag. But I don't think we can completely remove the hacky part of the code unless we are willing to break the backward compatibilities and the semantics.

**Comment by hkarel** [comment]
> The code above will cause a memory leak when your program creates a new thread and execute this path. Creating temporary, short-living threads on-demand at runtime is common, so I don't think it is an ideal solution.

I agree, I didn’t think about short-living threads (I don’t often use them).  
I'll change it to code with a flag, it's more reliable :))

> It's not a UB in this case. It's more like calling reset() or clear() function if GenericValue provide them.

That is, you want to say that if `GenericValue` is created using the default constructor, then in this case the `Allocator::Free(e)` function (see `~GenericValue()`) does not need to be called?

> But I don't think we can completely remove the hacky part of the code unless we are willing to break the backward compatibilities and the semantics.

With my question, I wanted to draw attention to the fact that even if the code being discussed does not lead to UB now, then in the future, if the behavior of the default constructor `GenericValue` changes - this can become an problem.

**Comment by jaepil** [comment]
Yes, in fact, `~GenericValue()` does not call `Allocator::Free()` if the value is null.

Suppose the default constructor has changed to allocate some resources dynamically in the future, which makes calling the destructor necessary. That means we will have a memory leak issue with this code with short-living threads.

However, as far as I know, the goal of this change was to keep the null-valued `GenericValue` alive during the termination, module unloading, or similar situations to prevent issues with object destruction order. So, any changes that make the destructor call necessary for null-valued `GenericValue` will cause a memory issue again. Your code is not an exception here. So I don't think the current implementation cause unnecessary problems. It's just a little bit hacky, and I agree with you that depending on the implementation details of the default constructor and destructor is not ideal.

I was thinking about a different approach regarding your suggestion of using a flag to prevent calling the constructor multiple times:

```c++
alignas(GenericValue) thread_local static char buffer[sizeof(GenericValue)];
thread_local static GenericValue* nullValue = new (buffer) GenericValue();
return *nullValue;
```

I will send a PR so you can review the change.

**Comment by jaepil** [comment]
It seems like some versions of MSVC cannot compile my solution. Also, I've found a potential issue with not calling the destructor-then-constructor each time. `GenericValue::operator[]()` returns non-const reference of the value. Since it's mutable, the user can change the value, and the change will be persistent. It changes the behavior of the `GenericValue::operator[]()` after that. Not only that, it gets worse if the user turns the value type into the array or object. In this case, we will have a memory leak.

It's the design of the interface causing this issue, but we cannot break the backward compatibilities, so the only option we have is to call destructor-then-constructor every time. Now I'm fully convinced that it needs to be addressed with more complex code. I will work on it in a few days.

**Comment by hkarel** [comment]
One more moment: in the old version MSVC (2008) I faced the following problem.
Code
```cpp
static int a = 5;
return a;
```
and 
```cpp
static int a (5);
return a;
```
have different behavior.  
In the first case, the variable `a` was initialized to 5 on every call. In the second case, variable `a` was initialized to 5 only on the first call.  
This issue has most likely been fixed by now.

**Comment by hkarel** [comment]
> So maybe inside GenericValue::operator[]() call these functions the explicitly on every call operator[]?

I saw `nullValue->SetNull()` :))
